### PR TITLE
Count unused classes in the Jenkins pipeline

### PR DESCRIPTION
**Pull request by sjoelund**

### CODE REVIEW

**Comment by dietmarw** [comment]
And us of ModelicaTest classes which is not really useful right? Easy to filter out but an unnecessary burden when using CI.

**Comment by sjoelund** [comment]
Well, yes. But if there's unused classes in ModelicaTest, those could just be removed... The same is also true for Examples, some classes in there seem unused (such as Modelica.Electrical.Analog.Examples.Utilities.RealSwitch)

**Comment by sjoelund** [comment]
Of course we could filter out some of these if we don't want to see them (Examples, UsersGuide, Icons, ModelicaTest, etc)

**Comment by dietmarw** [comment]
With "unused" you mean there are classes in ModelicaTest that do not have an experiment annotation? In that case, yes I agree they should be removed or/and reported.

**Comment by beutlich** [comment]
> The same is also true for Examples, some classes in there seem unused (such as Modelica.Electrical.Analog.Examples.Utilities.RealSwitch)

Yes, I already noticed. I will create a PR to get rid of them, but not for MSL v4.0.0.

**Comment by sjoelund** [comment]
> With "unused" you mean there are classes in ModelicaTest that do not have an experiment annotation? In that case, yes I agree they should be removed or/and reported.

You have for example ModelicaTest.Math.Random.automaticGlobalSeed, which is intended to test the global seed but not called from any experiment. There are 8 hits from other models that seem to call Modelica.Math.Random.Utilities.automaticGlobalSeed though, so it's probably not needed... The ModelicaTest version calls a function with a different signature, too. So it might even be wrong?

**Comment by HansOlsson** [comment]
I'm unsure if we should do this as part of CI since it isn't directly used to reject pull requests.
Having some form of number and graph of coverage over time could be useful - but that doesn't seem connected to CI, but more like a daily/weekly test.

**Comment by sjoelund** [comment]
I put it in the CI to not run during pull requests (only if the label I added for testing is used). So it will just be a few minutes of CPU-time every time there is an accepted PR in the background (and no waiting for checks to finish).

I think it's fine to run it on the master; you will always have the updated file available when needed without asking someone to regenerate the file or waiting for a weekly run. There are not many commits anyway (1 or 2 per day).

The usual case is to generate the libraries in the CI/CD system instead of putting them in the git like it is currently done; I think we should use CI/CD more.

**Comment by beutlich** [comment]
We should have it as CI job if we enforce full coverage now. This also means, we need to improve the script to not count Examples, Units, Icons etc.

File: .CI/Jenkinsfile [hunk 1]
```diff
@@ -8,15 +10,17 @@ pipeline {
...
         script {
-          def buildNumber = env.BUILD_NUMBER as int
-          if (buildNumber > 1) milestone(buildNumber - 1)
-          milestone(buildNumber)
+          /* If this is a change request, cancel previous builds if you push a new commit */
+          if (changeRequest()) {
+            def buildNumber = env.BUILD_NUMBER as int
+            if (buildNumber > 1) milestone(buildNumber - 1)
+            milestone(buildNumber)
+            doCountUnused = pullRequest.labels.contains("CI/Count unused classes")
```
Comment (beutlich) [inline .CI/Jenkinsfile]: Please change the label to "CI: class coverage".

File: .CI/CountClasses.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+allClasses = dict([(c.strip(),0) for c in open("all.classes","r").readlines() if c.strip()])
+for line in open("models.uses", "r").readlines():
+  line = line.strip().strip('"')
```
Comment (beutlich) [inline .CI/CountClasses.py]: Please move this file to a separate directory "classCoverage".

**Comment by sjoelund** [comment]
I fixed the comments.

**Pull request merged**